Syntax highlight for CodeBlock of OpenDocument

niszet opened the pull request:
By #6711, syntax highlighting for inline Code has been implemented.
Syntax highlighting for CodeBlock is also implemented based on Docx Writer.

* Paragraph style of CodeBlock is changed from "Preformatted_20_Text" to "Source_Code".
* As similar to docx writer, CodeBlock is decomposed to "Code + LineBreak" lines.
* To do this, inlineToOpenDocument Code is updated again to treat multiple lines.
* Styles will be incorporated by Writer/ODT.hs

Further discussion will be added to #6710.

niszet committed:
Syntax highlight for CodeBlock of OpenDocument

By #6711, syntax highlighting for inline Code has been implemented.
Syntax highlighting for CodeBlock of OpenDocument is also implemented based on Docx Writer.

* Paragraph style of CodeBlock is changed from "Preformatted_20_Text" to "Source_Code".
* As similar to docx writer, CodeBlock is decomposed to "Code + LineBreak" lines.
* To do this, inlineToOpenDocument for Code is updated again to treat multiple lines.
* Styles will be incorporated by Writer/ODT.hs

jgm commented on src/Text/Pandoc/Writers/OpenDocument.hs:
```diff
@@ -148,6 +151,11 @@ withTextStyle s f = do
 
+inSourceCodeTags :: PandocMonad m => Doc Text -> OD m (Doc Text)
+inSourceCodeTags d = do
+  modify $ \st -> st { stFirstPara = False }
```
Does the paragraph style you're using for this ensure that the first line is not indented?
(That was the main point of setting stFirstPara = True, so I just wanted to be sure.)

niszet replied:
I intended to disable `stFirstPara` but sorry this is my I misunderstanding. This Line (L156) will be removed in next commit.
Followings are memo for future who are reading this commit (maybe for me).

I saw `CodeBlock` calls [`Para` ](https://github.com/jgm/pandoc/blob/master/src/Text/Pandoc/Writers/Docx.hs#L985) and `Para` disables [`stFirstPara`](https://github.com/jgm/pandoc/blob/master/src/Text/Pandoc/Writers/Docx.hs#L970) .
But again, `CodeBlock` enables [`stFirstPara`](https://github.com/jgm/pandoc/blob/master/src/Text/Pandoc/Writers/Docx.hs#L986) . So, I will change the code as following.

```
    | CodeBlock    a s <- bs = setFirstPara >> (inlinesToOpenDocument o [Code a s] >>= inSourceCodeTags)
```

Basically, I would like to change the odt output to be the same as or similar to docx.

jgm commented:
OK, now I understand better than I did in your comment on the other issue.
I think this approach will work, yes!

niszet commented:
Thank you for reviewing. Basically, I will continue based on this concept.
But, entire of my plan to implement highlight is changed. I will write that concept to original issue #6710 (maybe tomorrow or later).

niszet commented:
Several Items of progress,

* I found a bug in this code. `CodeBlock` with `\n` is not converted to LineBreak correctly. This is fixed locally.
* I tried to update a test by updating [writer.opendocument](https://github.com/jgm/pandoc/blob/master/test/writer.opendocument). But original file cannot be opened by LibreOffice (LibreOffice says this xml has format error.). To check it, I placed this file into odt as a context.xml by my hand. I'm now checking grammatically incorrect portion and will update them together.

niszet committed:
Bug fix of syntax highlight for CodeBlock of OpenDocument

By #6717, syntax highlighting for  has been implemented.
But this commit has several bugs and this commit fixes these issues as followings.

* style-name of  is changed to "Source_20_Code" from "Source_Code".
* setFirstPara is added to CodeBlock. After CodeBlock, paragraph is treated as first paragraph.
* LineBreak is inserted to each lines of Code of CodeBlock.
*  is updated to pass test.
* To help reading, type annotations are added to several functions.

niszet commented:
Sorry for bothering. As I updated my comment above, original write.opendocument has no problem. 
My misunderstanding is caused by hand insertion to odt file. Filepath of figures are not same when I use `--to odt` and `--to opendocument`. (For odt output, filenames of figure are updated by odt writer and they are incorporated into its inside)
This includes update of `CodeBlock` and `writer.opendocument` for updating test.

Latest commit is excluding unintentionally incorporated local code for my test...

I will rebase and clean these commits and notice again.

jgm commented:
Can you rebase against current master so I can try this out?

tarleb commented:
I rebased this [here](https://github.com/tarleb/pandoc/tree/odt_highlight) and noted a small regression: it seems that code blocks that are nested in lists are no longer indented:

Current:
![orange-code-block](https://user-images.githubusercontent.com/507994/119169960-e7982000-ba62-11eb-9e5d-9d60d572c698.png)

odt_highlight:
![orange-code-block-new](https://user-images.githubusercontent.com/507994/119170019-f7176900-ba62-11eb-9c69-3a1b73c00bab.png)

@niszet, could you take a look?
